src/main/java/com/snowplowanalytics/snowplow/tracker/emitter/BatchEmitter.java

```diff
@@ -78,37 +86,93 @@ protected BatchEmitter(final Builder<?> builder) {
         Preconditions.checkArgument(builder.bufferSize > 0, "bufferSize must be greater than 0");
 
         this.bufferSize = builder.bufferSize;
+
+        bufferConsumer = new Thread(getBufferConsumerRunnable());
+        bufferConsumer.start();
     }
 
     /**
-     * Adds a payload to the buffer and checks whether we have reached the buffer
-     * limit yet.
+     * Adds a TrackerEvent to the concurrent queue buffer
      *
-     * @param payload an event payload
+     * @param event an event
      */
     @Override
-    public synchronized void emit(final TrackerPayload payload) {
-        buffer.add(payload);
-        if (buffer.size() >= bufferSize) {
-            flushBuffer();
+    public void emit(final TrackerEvent event) {
+        boolean result = eventBuffer.offer(event); // Add to buffer and quickly return back to application
+        
+        if (!result) {
+            LOGGER.error("Unable to add event to emitter, emitter buffer is full");
         }
     }
 
-    /**
-     * When the buffer limit is reached sending of the buffer is initiated.
+    /*
+     * Forces the events currently in the buffer to be sent
      */
+    @Override
     public void flushBuffer() {
-        execute(getRequestRunnable(buffer));
-        buffer = new ArrayList<>();
+        // Drain immediate event buffer
+        while (true) {
+            TrackerEvent event = eventBuffer.poll();
+            if (event == null) {
+                break;
+            } else {
+                eventsToSend.offer(event);
+            }
+        }
+
+        // Drain unsent events
+        List<TrackerEvent> events = new ArrayList<>();
+        eventsToSend.drainTo(events);
+
+        // And send them
+        execute(getRequestRunnable(events));
+    }
+
+    /**
+     * Returns List of Events that are in the buffer.
+     *
+     * @return the buffered events
+     */
+    @Override
+    public List<TrackerEvent> getBuffer() {
+        return eventsToSend.stream().collect(Collectors.toList());
+    }
+
+    /**
+     * Returns a Consumer for the concurrent queue buffer
+     * Consumes events onto another queue to be sent when bufferSize is reached
+     *
+     * @return the new Runnable object
+     */
+    private Runnable getBufferConsumerRunnable() {
+        return new Runnable() {
+            @Override
+            public void run() {
+                while (true) {
+                    try {
+                        eventsToSend.put(eventBuffer.take());
+                        if (eventsToSend.size() >= bufferSize) {
+                            List<TrackerEvent> events = new ArrayList<>();
+                            eventsToSend.drainTo(events);
+                            execute(getRequestRunnable(events));
```

Comment:
This little block is repeated in `flushBuffer`.  Just pointing that out in case you care about DRY.